File: backend/src/modules/usrpreferences/usrpreferences.service.ts
import { Injectable, NotFoundException } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository, Like } from 'typeorm';
import { UsrPreferences } from './entities/usr-preferences.entity';
import { PaginationDto, PaginatedResult } from '../../common/dto/pagination.dto';

@Injectable()
export class UsrpreferencesService {
  constructor(
    @InjectRepository(UsrPreferences)
    private usrPrefRepository: Repository<UsrPreferences>,
  ) {}

  async findAll(dto: PaginationDto): Promise<PaginatedResult<UsrPreferences>> {
    const { page, limit, keyword } = dto;
    const skip = (page - 1) * limit;
    const where: any = {};
    if (keyword) where.username = Like(`%${keyword}%`);
    const [items, total] = await this.usrPrefRepository.findAndCount({
      where, skip, take: limit, order: { id: 'DESC' },
    });
    return new PaginatedResult(items, total, page, limit);
  }

  async findByUser(username: string, domain: string): Promise<UsrPreferences[]> {
    return this.usrPrefRepository.find({
      where: { username, domain },
      order: { attribute: 'ASC' },
    });
  }

  async create(data: Partial<UsrPreferences>): Promise<UsrPreferences> {
    return this.usrPrefRepository.save(this.usrPrefRepository.create(data));
  }

  async update(id: number, data: Partial<UsrPreferences>): Promise<UsrPreferences> {
    const item = await this.usrPrefRepository.findOne({ where: { id } });
    if (!item) throw new NotFoundException('记录不存在');
    Object.assign(item, data);
    return this.usrPrefRepository.save(item);
  }

  async remove(id: number): Promise<void> {
    await this.usrPrefRepository.delete(id);
  }

  async removeByUser(username: string, domain: string): Promise<void> {
    await this.usrPrefRepository.delete({ username, domain });
  }
}

File: kam-admin-module-user/src/main/java/io/kamailio/admin/modules/usrpreferences/service/UsrpreferencesService.java
package io.kamailio.admin.modules.usrpreferences.service;

import io.kamailio.admin.common.dto.PaginatedResult;
import io.kamailio.admin.common.dto.PaginationDto;
import io.kamailio.admin.modules.usrpreferences.entity.UsrPreferences;
import io.kamailio.admin.modules.usrpreferences.repository.UsrPreferencesRepository;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
public class UsrpreferencesService {
    private final UsrPreferencesRepository repository;

    public UsrpreferencesService(UsrPreferencesRepository repository) {
        this.repository = repository;
    }

    public PaginatedResult<UsrPreferences> findAll(PaginationDto dto) {
        var page = dto.getKeyword() != null && !dto.getKeyword().isBlank()
                ? repository.findByUsernameContaining(dto.getKeyword(), PageRequest.of(dto.pageOrDefault() - 1, dto.limitOrDefault(), Sort.by(Sort.Direction.DESC, "id")))
                : repository.findAll(PageRequest.of(dto.pageOrDefault() - 1, dto.limitOrDefault(), Sort.by(Sort.Direction.DESC, "id")));
        return new PaginatedResult<>(page.getContent(), page.getTotalElements(), dto.pageOrDefault(), dto.limitOrDefault());
    }

    public List<UsrPreferences> findByUser(String username, String domain) {
        return repository.findByUsernameAndDomain(username, domain);
    }

    @Transactional
    public UsrPreferences create(UsrPreferences data) { return repository.save(data); }

    @Transactional
    public UsrPreferences update(Integer id, UsrPreferences data) {
        var item = repository.findById(id).orElseThrow(() -> new RuntimeException("记录不存在"));
        if (data.getUuid() != null) item.setUuid(data.getUuid());
        if (data.getUsername() != null) item.setUsername(data.getUsername());
        if (data.getDomain() != null) item.setDomain(data.getDomain());
        if (data.getAttribute() != null) item.setAttribute(data.getAttribute());
        if (data.getType() != null) item.setType(data.getType());
        if (data.getValue() != null) item.setValue(data.getValue());
        return repository.save(item);
    }

    @Transactional
    public void remove(Integer id) { repository.deleteById(id); }

    @Transactional
    public void removeByUser(String username, String domain) {
        repository.deleteByUsernameAndDomain(username, domain);
    }
}
